src/emulation_core/mips/datapath.rs

```diff
@@ -230,37 +237,135 @@ impl MipsDatapath {
+                self.state.rt = i.rt as u32;
+            }
+            Instruction::PlaceholderType(p) => {
+                self.state.rs = p.rs as u32;
+                self.state.rt = p.rt as u32;
+            }
+            _ => panic!("fixme"),
+        }
+
+        match &self.instruction_enum {
+            Instruction::RType(r) => {
+                self.state.shamt = r.shamt as u32;
+                self.state.funct = r.funct as u32;
+            }
+            _ => {
+                // placholder for non-R-type
+                self.state.shamt = 0;
+                self.state.funct = 0;
+            }
+        }
+
+        match &self.instruction_enum {
+            Instruction::RType(r) => {
+                self.state.rs = r.rs as u32;
+                self.state.rt = r.rt as u32;
+                self.state.rd = r.rd as u32;
+            }
+            Instruction::IType(i) => {
+                self.state.rs = i.rs as u32;
+                self.state.rt = i.rt as u32;
+                self.state.rd = 0; // Placeholder
+            }
+            Instruction::PlaceholderType(p) => {
+                self.state.rs = p.rs as u32;
+                self.state.rt = p.rt as u32;
+                self.state.rd = p.rd as u32;
+            }
+            _ => panic!("fixme"),
+        }
     }
 
     /// Extend the sign of a 16-bit value to the other 48 bits of a
     /// 64-bit value.
     fn sign_extend(&mut self) {
-        // Is the value negative or positive? Check sign bit
-
-        // 0000 0000 0000 0000 1000 0000 0000 0000
-        // 0x00008000
+        self.state.imm = ((self.state.imm as i16) as i32) as u32;
+    }
 
-        self.sign_extend = if (self.imm & 0x00008000) >> 15 == 0 {
-            self.imm as u64
-        } else {
-            (self.imm as u64) | 0xFFFF_FFFF_FFFF_0000
+    fn set_itype_control_signals(&mut self, i: IType) {
+        match i.op {
+            // Or immediate (ori)
+            // this code really needs to be moved somewhere else
+            0b001101 => {
+                self.signals.alu_op = AluOp::Or;
+                self.signals.alu_src = AluSrc::ZeroExtendedImmediate;
+                self.signals.branch = Branch::NoBranch;
+                self.signals.imm_shift = ImmShift::Shift0;
+                self.signals.jump = Jump::NoJump;
+                self.signals.mem_read = MemRead::NoRead;
+                self.signals.mem_to_reg = MemToReg::UseAlu;
+                self.signals.mem_write = MemWrite::NoWrite;
+                self.signals.mem_write_src = MemWriteSrc::PrimaryUnit;
+                self.signals.reg_dst = RegDst::Reg2;
+                self.signals.reg_width = RegWidth::DoubleWord;
+                self.signals.reg_write = RegWrite::YesWrite;
+            }
+            _ => panic!("Unsupported itype instructions"),
         }
     }
 
     /// Set the control signals for the datapath based on the
     /// instruction's opcode.
     fn set_control_signals(&mut self) {
```

Comment:
This function in general seems to still have some questions. I would like to talk about this some more about how we want to set the control signals.